### PR TITLE
Added ViewModelBase that implements INotifyPropertyChanged

### PR DESCRIPTION
**Pull request by gianpaolof**
Added methods to avoid backing fields in the get/set

---
Name: Refactoring
---
I created a base class for all the view models in order to centralize the OnPropertyChange calls
Plus I have added some methods that lets you write accessor without backing fields (it hits performance but for this application I guess you may accept it)

### CODE REVIEW

**Comment by 0xaryan** [comment]
Thank you @gianpaolof for your contribution! I'll check it out asap.

**Comment by gianpaolof** [comment]
You are welcome. Hope you can find it useful. Very nice app, congrats

Il mer 21 ago 2019, 18:24 Aryan Ebrahimpour <notifications@github.com> ha
scritto:

File: Source/ViewModels/HeaderMenuViewModel.cs [hunk 1]
```diff
@@ -1,56 +1,32 @@
...
         public bool IsTimeItemChecked
         {
-            get { return _isTimeItemChecked; }
-            set
-            {
-                _isTimeItemChecked = value;
-                OnPropertyChanged();
-            }
+            get { return GetValue(() => IsTimeItemChecked); }
+            set { SetValue(() => IsTimeItemChecked, value); }
```
Comment (0xaryan) [inline Source/ViewModels/HeaderMenuViewModel.cs]: Can you also simplify getter/setters with expression-bodied functions?
```csharp
public bool IsTimeItemChecked {
     get => GetValue(() => IsTimeItemChecked);
     set => SetValue(() => IsTimeItemChecked, value);
}
```

File: Source/ViewModels/ViewModelBase.cs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+using System.Diagnostics;
+using System.Linq.Expressions;
+using System.Runtime.CompilerServices;
+
+namespace EyesGuard.ViewModels
+{
+    /// <summary>
+    /// Base class for all the viewmodels. 
+    /// Implements INotifyPropertyChanged
+    /// main idea from http://web.archive.org/web/20121105051845/http://dotnet-forum.de/blogs/thearchitect/archive/2012/11/01/die-optimale-implementierung-des-inotifypropertychanged-interfaces.aspx
```
Comment (0xaryan) [inline Source/ViewModels/ViewModelBase.cs]: I think it would be better if we put "Main idea from ..." in `<remarks>` section instead of `<summary>`

**Comment by 0xaryan** [comment]
> Hope you can find it useful.

It is very useful for sure.

> Very nice app, congrats

Thank you @gianpaolof (:

**Comment by 0xaryan** [comment]
As I want to make some changes tonight, I'll take care of change requests.

Thank you!

**Comment by gianpaolof** [comment]
Hello!
I'll update the code as per your suggestion, tomorrow,after a synch. Glad
it helps.

Il mer 21 ago 2019, 21:32 Aryan Ebrahimpour <notifications@github.com> ha
scritto:

**Comment by 0xaryan** [comment]
Well, there will be a HUGE change in the codebase. I want to restructure the project to use F# Json Type provider for project's translation, and Paket as the package manager.

**Comment by gianpaolof** [comment]
You merged it already, cool.

I'll synch tomorrow, curious to see F# stuff

Il mer 21 ago 2019, 21:49 Aryan Ebrahimpour <notifications@github.com> ha
scritto:

**Pull request merged**